src/vector/focus.rs

```diff
@@ -748,18 +772,33 @@ where
     }
 }
 
+// NOTE: The documentation the mutable version is similar to the non-mutable version. I will comment for the places
+// where there are differences, otherwise the documentation is copied directly.
 pub struct TreeFocusMut<'a, A> {
+    /// The tree that this TreeFocusMut refers to. Unlike the non-mutable version, TreeFocusMut needs to store a
+    /// mutable reference. Additionally, there may be multiple TreeFocusMuts that refer to the same tree so we need a
+    /// Lock to synchronise the changes.
     tree: Lock<&'a mut RRB<A>>,
+    /// The view represents the range of the tree that this TreeFocusMut can see. The view can be narrowed by calling
+    /// either the narrow or split_at methods.
     view: Range<usize>,
+    /// The tree version of the Vector is represented as the concatenation of 2 chunks, followed by the tree root,
+    /// followed by 2 chunks. The middle_range refers to the range of the Vector that the tree covers.
     middle_range: Range<usize>,
+    /// This implementation of a focusmut stores only a single chunk for the Vector. This chunk can refer to one of the
+    /// 4 chunks front/back chunks or one of the leaves of the tree. The target_ptr is the pointer to the actual chunk
+    /// in question. The target_range is the range that the chunk represents.
     target_range: Range<usize>,
+    /// Not actually sure why this needs to be an atomic, it seems like it is unneccessary. This is just a pointer to
+    /// the chunk referred to above.
```

Comment:
I don't think AtomicPtr is necessary here - every where the chunk pointer is set is a &mut self method.

Reply:
Yes, but you can have multiple `FocusMut`s on different threads with the same chunk pointer, right? Because if you do a split that goes down the middle of a chunk, then both parts could be pointing at that chunk. Now, all the writes are in a section protected by a `Mutex`, but I don't actually know off the top of my head to what extent memory writes in a section protected by a mutex are prevented from re-ordering. Maybe they need to be atomic for that?

Reply:
If you have two FocusMuts that are both sharing a chunk they will each have their own AtomicPtr instead of one shared one. I had to search for exactly what AtomicPtr protects as the docs are a bit sparse. I found some help here: http://blog.rustbr.org/en/understanding-atomics/ . It just seems to just be the memory address itself, which I don't think needs protection here.

Reply:
Oh right, I was totally making the mistake of thinking that the atomic thing is the write through the pointer...